refactor: AQMFile class

pcolaianni opened the pull request:
It's supposed to fix that problem you had with empty cells in params headers.

I also went into a refactor of the class, hoping it looks cleaner now.

Please let me know if it works well :smile: 

NOTE: I'm assuming that some headers (all but the custom params' headers) are present. Not their values, but just the headers. Any value (cell?) can be empty. But the headers could not (again, not counting the custom params like `slope` etc)

pcolaianni committed:
refactor: AQMFile class
fix: support empty cells in params

pcolaianni committed:
style: move code, easier diff

dmccloskey commented on src/openms/source/FORMAT/AbsoluteQuantitationMethodFile.cpp:
```diff
@@ -38,176 +38,77 @@ namespace OpenMS
+        headers[sl[i]] = i; // for each header found, assign an index value to it
       }
-      else
+      if (!( // if any of these headers is missing, exception is thrown
```
I am not sure if it is needed to throw an exception if a header is not found.  The code should provide a default value if the header is missing.  However, it might be useful for the reader to know if they missed a header if it was done by accident.

pcolaianni replied:
Ok, so I'll give an empty string `""` to members of type `String` and `0` to numeric members. I'll look for a way to notify the user.

dmccloskey commented on src/openms/source/FORMAT/AbsoluteQuantitationMethodFile.cpp:
```diff
@@ -38,176 +38,77 @@ namespace OpenMS
-        transformation_model_params.setValue(kv.first,std::stoi(line[kv.second]));
-      }
-      else
+    StringList tl = line; // trimmed line
```
Good idea trimming out the white spaces.  Does this method remove only the starting and ending white spaces or all whitespaces even between words?  If the later, that may be an issue because a component could be named e.g., "Glucose 6 phosphate" and in that case the spaces between the words would be important.

pcolaianni replied:
It removes only whitespaces on the boundaries. The spaces within the name are not removed.

dmccloskey commented on src/openms/source/FORMAT/AbsoluteQuantitationMethodFile.cpp:
```diff
@@ -38,176 +38,77 @@ namespace OpenMS
+    {
+      s.trim();
+    }
+    aqm.setComponentName(tl[headers.at("component_name")]);
```
Nice reduction in the amount of lines.

dmccloskey commented on src/openms/source/FORMAT/AbsoluteQuantitationMethodFile.cpp:
```diff
@@ -253,10 +148,34 @@ namespace OpenMS
       for (Size i = 0, j = 11; i < tm_params_names.size(); ++i, ++j)
       {
-        row[j] = tm_params.getValue(tm_params_names[i]);
+        row[j] = tm_params.exists(tm_params_names[i]) ? tm_params.getValue(tm_params_names[i]) : "";
```
Does your unit tests pass with this line (e.g., if the parameter is missing)?  I would imagine if the type of the param is not compatible with an string it would throw an error.

pcolaianni replied:
That's a good point. The tests pass but only because I'm testing that input and output files look the same.
I guess I should be checking the param name and lookup its type, and give a "default" value of `""` or `0` according to the found type.

dmccloskey commented on src/tests/class_tests/openms/source/AbsoluteQuantitationMethodFile_test.cpp:
```diff
@@ -284,8 +217,13 @@ START_SECTION(void store(const String & filename, const std::vector<AbsoluteQuan
     const Param& tm_params1 = aqm_list1[i].getTransformationModelParams();
     const Param& tm_params2 = aqm_list2[i].getTransformationModelParams();
     TEST_EQUAL(tm_params1.size(), tm_params2.size())
-    TEST_REAL_SIMILAR(tm_params1.getValue("slope"), tm_params2.getValue("slope"))
```
It is tough to tell from the diff, but does this test for the initial problem that we were discussing?

pcolaianni replied:
Given headers eg. `slope,intercept` etc (imagine the full names, not just the suffixes), the code does not assume that the cell will contain data. Instead, it uses a default value for the given member's type. This was the original problem.
I added some new lines in the .csv file (1 line misses a param's value, and others miss other information in other columns) to make sure no crash happens now.

dmccloskey reviewed (commented):
The changes definitely decreased the length of the code.  The use of boost.regex to id the transformation_model_params_ headers was nice.  I was concerned about a few of the updates introducing an unintended side effect.

pcolaianni committed:
refactor: Support missing headers

Warn the user instead of throw an exception

Adjust the logic to support missing headers

Set a default value also for empty params cells, instead of skipping them

Improve tests

pcolaianni commented:
PR updated, also this issue https://github.com/biosustain/OpenMS/issues/111 should be fixed with the last commit.
@dmccloskey

dmccloskey commented on src/openms/source/FORMAT/AbsoluteQuantitationMethodFile.cpp:
```diff
@@ -63,7 +63,18 @@ namespace OpenMS
         headers.count("transformation_model")
       ))
       {
-        throw Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "Required columns are missing.");
```
This should be much more user friendly.

dmccloskey commented on src/openms/source/FORMAT/AbsoluteQuantitationMethodFile.cpp:
```diff
@@ -86,17 +97,21 @@ namespace OpenMS
+    aqm.setComponentName(headers.count("component_name") ? tl[headers.at("component_name")] : "");
+    aqm.setFeatureName(headers.count("feature_name") ? tl[headers.at("feature_name")] : "");
+    aqm.setISName(headers.count("IS_name") ? tl[headers.at("IS_name")] : "");
+    aqm.setLLOD(!headers.count("llod") || tl[headers.at("llod")].empty() ? 0 : std::stod(tl[headers.at("llod")]));
```
OK this is the behavior the original code was trying to capture...except in 1 line instead of 5...which is nice.

dmccloskey commented on src/tests/class_tests/openms/source/AbsoluteQuantitationMethodFile_test.cpp:
```diff
@@ -192,14 +208,62 @@ START_SECTION(void load(const String & filename, std::vector<AbsoluteQuantitatio
   TEST_REAL_SIMILAR(transformation_model_params.getValue("slope"), 1.0);
   TEST_REAL_SIMILAR(transformation_model_params.getValue("intercept"), 2.0);
+
+  // The following input file doesn't have the headers: component_name, llod
```
Are you checking for a missing `transformation_model_param_` column as well?

pcolaianni replied:
Yes, for example here: https://github.com/biosustain/OpenMS/pull/110/files/96dc372ab4cedc953abb3beecd66325ca98566c6..364567b3d1d0265751ca48019f5b2562b2e2d332#diff-966d9b27049c1f1c5958431fecea06c7R242
No value was present within the file, so a default value is set.

Same at line 194.
